#Emotion Classification from Text using Machine Learning

In this project, machine learning models are developed to classify emotions from textual data. Emotion classification is an important task in Natural Language Processing (NLP) and is widely used in applications such as sentiment analysis, social media monitoring, and human–computer interaction.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv('/content/nlp_dataset.csv')
df

,Comment,Emotion
0,i seriously hate one subject to death but now ...,fear
1,im so full of life i feel appalled,anger
2,i sit here to write i start to dig out my feel...,fear
3,ive been really angry with r and i feel like a...,joy
4,i feel suspicious if there is no one outside l...,fear
...,...,...
5932,i begun to feel distressed for you,fear
5933,i left feeling annoyed and angry thinking that...,anger
5934,i were to ever get married i d have everything...,joy
5935,i feel reluctant in applying there because i w...,fear


In [3]:
df['Emotion'].value_counts()

,count
Emotion,
anger,2000
joy,2000
fear,1937


In [4]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
df['Emotion']=le.fit_transform(df['Emotion'])
df

,Comment,Emotion
0,i seriously hate one subject to death but now ...,1
1,im so full of life i feel appalled,0
2,i sit here to write i start to dig out my feel...,1
3,ive been really angry with r and i feel like a...,2
4,i feel suspicious if there is no one outside l...,1
...,...,...
5932,i begun to feel distressed for you,1
5933,i left feeling annoyed and angry thinking that...,0
5934,i were to ever get married i d have everything...,2
5935,i feel reluctant in applying there because i w...,1


In [7]:
comment=df.Comment
comment

,Comment
0,i seriously hate one subject to death but now ...
1,im so full of life i feel appalled
2,i sit here to write i start to dig out my feel...
3,ive been really angry with r and i feel like a...
4,i feel suspicious if there is no one outside l...
...,...
5932,i begun to feel distressed for you
5933,i left feeling annoyed and angry thinking that...
5934,i were to ever get married i d have everything...
5935,i feel reluctant in applying there because i w...


In [10]:
comment=comment.str.replace ("[^a-zA-Z0-9]+"," ",regex=True)
comment

,Comment
0,i seriously hate one subject to death but now ...
1,im so full of life i feel appalled
2,i sit here to write i start to dig out my feel...
3,ive been really angry with r and i feel like a...
4,i feel suspicious if there is no one outside l...
...,...
5932,i begun to feel distressed for you
5933,i left feeling annoyed and angry thinking that...
5934,i were to ever get married i d have everything...
5935,i feel reluctant in applying there because i w...


In [16]:
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from nltk.corpus import wordnet
from nltk import pos_tag

nltk.download('punkt')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [17]:
def get_wordnet_pos(word):
    tag = pos_tag([word])[0][1][0].upper()
    tag_dict = {
        "J": wordnet.ADJ,
        "N": wordnet.NOUN,
        "V": wordnet.VERB,
        "R": wordnet.ADV
    }
    return tag_dict.get(tag, wordnet.NOUN)

In [18]:
lemmatizer = WordNetLemmatizer()

def lemmatize_text(comment):
    tokens = word_tokenize(str(comment))
    lemmas = [lemmatizer.lemmatize(word, get_wordnet_pos(word)) for word in tokens]
    return " ".join(lemmas)

In [20]:
nltk.download('averaged_perceptron_tagger_eng')
comment_lemma = comment.apply(lemmatize_text)

print(comment_lemma.head())

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger_eng.zip.


0    i seriously hate one subject to death but now ...
1                   im so full of life i feel appalled
2    i sit here to write i start to dig out my feel...
3    ive be really angry with r and i feel like an ...
4    i feel suspicious if there be no one outside l...
Name: Comment, dtype: object


Lemmatization is applied to the text data to convert words into their base or dictionary form using NLTK’s WordNetLemmatizer. Each comment is first tokenized into individual words, and then each word is reduced to its root form (lemma). For example, words like “playing”, “studies”, and “cars” are converted to “play”, “study”, and “car”. This step helps reduce vocabulary size and ensures that different forms of the same word are treated as a single feature, improving the efficiency and performance of machine learning models.

In [28]:
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')
sw=stopwords.words("english")
#sw

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Stopword removal is performed to eliminate commonly used words such as “the”, “is”, and “and” that do not contribute significant meaning to the text. Removing these words reduces noise in the dataset and helps the machine learning model focus on more informative words that are useful for emotion classification

In [26]:
from nltk import TweetTokenizer
tk=TweetTokenizer()


In [27]:
comment=comment.apply (lambda x:[i for i in tk.tokenize(x) if i not in sw]).apply (lambda i:" ".join(i))
comment

,Comment
0,seriously hate one subject death feel reluctan...
1,im full life feel appalled
2,sit write start dig feelings think afraid acce...
3,ive really angry r feel like idiot trusting fi...
4,feel suspicious one outside like rapture happe...
...,...
5932,begun feel distressed
5933,left feeling annoyed angry thinking center stu...
5934,ever get married everything ready offer got to...
5935,feel reluctant applying want able find company...


TweetTokenizer from the NLTK library is used to tokenize the text into individual words. During tokenization, stopwords such as “the”, “is”, and “and” are removed to eliminate commonly occurring words that do not add meaningful information. After removing stopwords, the remaining tokens are joined back into a cleaned text format. This step helps reduce noise in the data and improves the quality of features used for machine learning models.

In [30]:
#vectorization
from sklearn.feature_extraction.text import TfidfVectorizer
vector=TfidfVectorizer()
X=vector.fit_transform(comment)
print(X)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 53883 stored elements and shape (5937, 8817)>
  Coords	Values
  (0, 6791)	0.38803028564104103
  (0, 3519)	0.33010756312888306
  (0, 5333)	0.252897126981499
  (0, 7481)	0.44547688485006304
  (0, 1880)	0.41947362137972544
  (0, 2855)	0.08263907638284912
  (0, 6336)	0.3218747945164998
  (0, 2319)	0.4377946585961953
  (1, 2855)	0.1279614344273125
  (1, 3813)	0.26426606619168347
  (1, 3142)	0.5728988776878813
  (1, 4462)	0.40458466198963866
  (1, 359)	0.6495286327680816
  (2, 6986)	0.312901511326718
  (2, 8726)	0.2714542333503176
  (2, 7325)	0.2769159531198503
  (2, 2091)	0.41095194523172573
  (2, 2860)	0.2506523025993543
  (2, 7817)	0.21538360913503146
  (2, 161)	0.2652157991200991
  (2, 36)	0.36559978105852053
  (2, 5813)	0.3766443574373461
  (2, 4848)	0.27859392353233836
  (2, 4656)	0.2239340414268402
  (3, 2855)	0.0932141825140457
  :	:
  (5934, 6748)	0.21832652375534806
  (5934, 4565)	0.21395576602050526
  (5934, 4709)	0.280

TF-IDF Vectorization is used to convert the cleaned text data into numerical features that can be used by machine learning models. TF-IDF assigns weights to words based on their importance in a document relative to the entire dataset. Words that appear frequently in a specific text but rarely across other texts receive higher weights. The resulting output is a sparse matrix where rows represent text samples and columns represent unique words in the vocabulary.

In [32]:
y=df['Emotion']


In [33]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [34]:
from sklearn.naive_bayes import MultinomialNB
nb=MultinomialNB()
nb.fit(X_train,y_train)


MultinomialNB()

In [35]:
y_pred_nb=nb.predict(X_test)
y_pred_nb

array([0, 2, 1, ..., 0, 1, 2])

In [36]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print(accuracy_score(y_test,y_pred_nb))
print(classification_report(y_test,y_pred_nb))
print(confusion_matrix(y_test,y_pred_nb))

0.9107744107744108
              precision    recall  f1-score   support

           0       0.88      0.95      0.91       392
           1       0.92      0.92      0.92       416
           2       0.94      0.87      0.90       380

    accuracy                           0.91      1188
   macro avg       0.91      0.91      0.91      1188
weighted avg       0.91      0.91      0.91      1188

[[371  12   9]
 [ 23 382  11]
 [ 29  22 329]]


#Naive Bayes Model Evaluation

The Multinomial Naive Bayes model was trained using TF-IDF features extracted from the preprocessed text data. The dataset was split into training and testing sets, and the model's performance was evaluated using accuracy, precision, recall, and F1-score.

The model achieved an accuracy of approximately 91%, indicating that it correctly classified most of the text samples into their respective emotion categories.

#Classification Metrics

Precision measures how many of the predicted labels are actually correct.

Recall measures how many of the actual labels are correctly identified by the model.

F1-score is the harmonic mean of precision and recall, providing a balanced measure of model performance.

The results show that the model performs consistently across all emotion classes with F1-scores around 0.90–0.92, indicating balanced performance.

Confusion Matrix

The confusion matrix shows the number of correct and incorrect predictions for each class. Most values appear along the diagonal, which indicates that the model correctly classified the majority of samples.

In [37]:
from sklearn.svm import SVC
svm=SVC()
svm.fit(X_train,y_train)

SVC()

In [38]:
y_pred_svc=svm.predict(X_test)
y_pred_svc

array([0, 2, 1, ..., 2, 1, 2])

In [39]:
print(accuracy_score(y_test,y_pred_svc))
print(classification_report(y_test,y_pred_svc))
print(confusion_matrix(y_test,y_pred_svc))

0.9343434343434344
              precision    recall  f1-score   support

           0       0.93      0.93      0.93       392
           1       0.97      0.90      0.93       416
           2       0.90      0.97      0.94       380

    accuracy                           0.93      1188
   macro avg       0.94      0.94      0.93      1188
weighted avg       0.94      0.93      0.93      1188

[[366   5  21]
 [ 22 374  20]
 [  4   6 370]]


#Support Vector Machine (SVM) Model Evaluation

A Support Vector Machine (SVM) classifier was trained using the TF-IDF features extracted from the preprocessed text data. SVM is a powerful supervised learning algorithm that finds the optimal boundary (hyperplane) to separate different classes. It is particularly effective for high-dimensional data such as text classification.

The model achieved an accuracy of approximately 93.4%, which is slightly higher than the Naive Bayes model. The precision, recall, and F1-scores for all classes are around 0.93–0.94, indicating that the model performs well across all emotion categories.

#Confusion Matrix Interpretation

The confusion matrix shows that most predictions lie along the diagonal, meaning that the majority of samples were correctly classified. Only a small number of samples were misclassified between emotion classes, demonstrating that the SVM model provides strong performance for emotion classification.

#Model Comparison
The performance of Naive Bayes and SVM models was compared using accuracy and F1-score. The Naive Bayes model achieved an accuracy of 91%, while the SVM model achieved 93.4% accuracy. This indicates that SVM performs slightly better for this emotion classification task. However, Naive Bayes is computationally simpler and faster, while SVM provides better classification boundaries for complex datasets.